File: src/core/git.ts
import { spawnSync } from "node:child_process";
import path from "node:path";
import process from "node:process";

export function gitOutput(repoPath: string, ...args: string[]): string {
  const completed = spawnSync("git", ["-C", repoPath, ...args], {
    encoding: "utf8",
    stdio: ["ignore", "pipe", "pipe"],
  });
  if (completed.status !== 0) {
    throw new Error((completed.stderr || completed.stdout || "git command failed").trim());
  }
  return completed.stdout.trim();
}

export function inferRepoRef(repoPath: string): string {
  const remoteUrl = gitOutput(repoPath, "remote", "get-url", "origin").replace(/\.git$/, "");
  for (const prefix of ["git@github.com:", "https://github.com/", "http://github.com/"]) {
    if (remoteUrl.startsWith(prefix)) return remoteUrl.slice(prefix.length);
  }
  return remoteUrl;
}

export function inferDefaultBranch(repoPath: string): string {
  try {
    return gitOutput(repoPath, "branch", "--show-current") || "main";
  } catch {
    return "main";
  }
}

export function findRepoRoot(startPath = process.cwd()): string | null {
  try {
    return gitOutput(startPath, "rev-parse", "--show-toplevel");
  } catch {
    return null;
  }
}

export function resolveRepoPath(repoPath?: string): string | null {
  if (repoPath) return path.resolve(repoPath);
  return findRepoRoot();
}

export function gitDiff(repoPath: string, base?: string): string {
  try {
    if (base) return gitOutput(repoPath, "diff", "--name-only", base);
    return gitOutput(repoPath, "diff", "--name-only", "HEAD");
  } catch {
    return "";
  }
}

export function gitChangedFiles(repoPath: string): string[] {
  const diff = gitDiff(repoPath);
  return diff ? diff.split("\n").filter(Boolean) : [];
}

File: dops/git.py
from __future__ import annotations

import os
import subprocess
from pathlib import Path


def git_output(repo_path: str, *args: str) -> str:
    completed = subprocess.run(
        ["git", "-C", repo_path, *args],
        capture_output=True,
        text=True,
        check=False,
    )
    if completed.returncode != 0:
        raise RuntimeError((completed.stderr or completed.stdout or "git command failed").strip())
    return completed.stdout.strip()


def infer_repo_ref(repo_path: str) -> str:
    remote_url = git_output(repo_path, "remote", "get-url", "origin").removesuffix(".git")
    for prefix in ("git@github.com:", "https://github.com/", "http://github.com/"):
        if remote_url.startswith(prefix):
            return remote_url[len(prefix) :]
    return remote_url


def infer_default_branch(repo_path: str) -> str:
    try:
        return git_output(repo_path, "branch", "--show-current") or "main"
    except RuntimeError:
        return "main"


def find_repo_root(start_path: str | None = None) -> str | None:
    try:
        return git_output(start_path or os.getcwd(), "rev-parse", "--show-toplevel")
    except RuntimeError:
        return None


def resolve_repo_path(repo_path: str | None = None) -> str | None:
    if repo_path:
        return str(Path(repo_path).resolve())
    return find_repo_root()


def git_diff(repo_path: str, base: str | None = None) -> str:
    try:
        if base:
            return git_output(repo_path, "diff", "--name-only", base)
        return git_output(repo_path, "diff", "--name-only", "HEAD")
    except RuntimeError:
        return ""


def git_changed_files(repo_path: str) -> list[str]:
    diff = git_diff(repo_path)
    return [line for line in diff.splitlines() if line]
